cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -265,5 +265,47 @@ class ARROW_EXPORT SubTreeFileSystem : public FileSystem {
   Status FixStats(FileStats* st) const;
 };
 
+/// \brief EXPERIMENTAL: a FileSystem implementation that delegates to another
+/// implementation but inserts latencies at various points.
+class ARROW_EXPORT SlowFileSystem : public FileSystem {
```

Comment:
I've brought this topic up before, but what do you think about putting the entire implementation of SlowFS in a .cc file and returning std::shared_ptr<FileSystem> from the function that creates it? This can always be done later so refactor need not happen now. It's probably better to use factory methods for instantiating most FS classes anyway

Unless you anticipate adding additional methods to the class

Reply:
Same comments hold regarding the slow stream classes

Reply:
My preference would be to hide the implementations in all cases except where we expose additional methods. If only that it offers more freedom to refactor

Reply:
Ok, we discuss this in a separate JIRA. Will merge.